text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+## Generic parameter
+
+Change the trait definition to have a generic parameter instead of an associated type. There might be permutations here which could allow a single smart pointer type to dispatch method calls to multiple possible receivers - but this would add complexity, no known use case exists, and it might cause worst-case O(n^2) performance on method lookup.
+
+## Do not enable for pointers
+
+It would be possible to respect the `Receiver` trait without allowing dispatch onto raw pointers - they are essentially independent changes to the candidate deduction algorithm.
+
+We don't want to encourage the use of raw pointers, and would prefer rather that raw pointers are wrapped in a custom smart pointer that encodes and documents the invariants. So, there's an argument not to add the raw pointer support.
+
+However, the current unstable `arbitrary_self_types` feature provides support for raw pointer receivers, and with years of experience no major concerns have been spotted. We would prefer not to deviate from the existing proposal more than necessary. Moreover, we are led to believe that raw pointer receivers are quite important for the future of safe Rust, because stacked borrows makes it illegal to materialize references in many positions, and there are a lot of operations (like going from a raw pointer to a raw pointer to a field) where users don't need to or want to do that. We think the utility of including raw pointer receivers outweighs the risks of tempting people to over-use raw pointers.
+
+## Provide compiler support for dereferencing pointers
+
+This RFC proposes to implement `Receiver` for `*mut T` and `*const T` within the library. This is slightly different from the unstable arbitrary self types support, which instead hard-codes pointer support into the candidate deduction algorithm in the compiler (because obviously `Deref` can't be implemented for pointers.)
+
+We prefer the option of specifying behavior in the library using the normal trait, though it's a compatibility break for users of Rust who don't adopt the `core` crate (including compiler tests).
+
+## Implement for `Weak` and `NonNull`
+
+`Weak<T>` and `NonNull<T>` were not supported by the prior unstable arbitrary self types support, but they share the property that it may be desirable to implement method calls to `T` using them as self types. Unfortunately they also share the property that these types have many Rust methods using `self`, `&self` or `&mut self`. If we added to the set of Rust methods in future, we'd [shadow any such method calls](#method-shadowing). We can't implement `Receiver` for these types unless we come up with a policy that all subsequent additions to these types would instead be associated functions. That would make the future APIs for these types a confusing mismash of methods and associated functions, and the extra user complexity doesn't seem merited.
```

Comment:
This applies to `*const T` too; it has [many methods](https://doc.rust-lang.org/std/primitive.pointer.html). Allowing methods that take `*const Self` seems pretty important to this RFC, maybe we can tweak the resolution algorithm instead? I'm not sure what's a good solution here.

Reply:
One option would be that method resolution that goes through `Receiver` has higher priority than inherent methods on the pointer type. I pointed out in my other comment that `Deref` still needs a special-case in the resolution algorithm, so we can keep the current priority for these. This would be pretty confusing to use tho.

We could even special-case raw ptrs, `NonNull` and the like in method resolution :')

Reply:
I think it would probably be ideal to have a dedicated diagnostic for shadowing built-in pointer methods like this, if we go the route of adding pointer receivers. People would expect `addr` to work normally, and overriding it yourself seems like a bad idea.